# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1: Programming

Note: the exercise numbers below match up with the reading for ease of cross referencing.

**high-level languages** -- for humans

    ex) C, C++, Java, Python, Perl 

    * portable, shorter, easier to read, processed

**low-level languages** -- for computer

    = "machine language" = "assembly language" 

high-level --> low-level
* interpreters = source code --> interpreter --> output
* compilers = source code --> compiler --> object code --> executor --> output

python = interpreter
1. interactive mode
  * ">>>" = prompt
2. script mode
  * stores code in a file (__.py) --> call name

**PROGRAM** = sequence of instructions --> perform computation
   input | output | math | conditional execution
   
**BUGS** = errors
1. syntax
2. runtime 
3. semantic

python = formal language = tokens + structure
*  parsing = determine structure of sentence 

### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [1]:
help('print')

The ``print`` statement
***********************

   print_stmt ::= "print" ([expression ("," expression)* [","]]
                  | ">>" expression [("," expression)+ [","]])

``print`` evaluates each expression in turn and writes the resulting
object to standard output (see below).  If an object is not a string,
it is first converted to a string using the rules for string
conversions.  The (resulting or original) string is then written.  A
space is written before each object is (converted and) written, unless
the output system believes it is positioned at the beginning of a
line.  This is the case (1) when no characters have yet been written
to standard output, (2) when the last character written to standard
output is a whitespace character except ``' '``, or (3) when the last
write operation on standard output was not a ``print`` statement. (In
some cases it may be functional to write an empty string to standard
output for this reason.)

Note: Objects which act like file objects but

### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 

In [7]:
((43 * 60 + 30) / (10 / 1.61)) / 3600

0.11672500000000001

0.117 hr per mile

In [7]:
(10 * 3600) / ((43 * 60 + 30) * 1.61)

8.567144998929106

8.57 mph

REVISION: write more code structure than calculations

In [4]:
def python_calc():
    time_sec = 43.0 * 60.0 + 30.0
    time_hr = time_sec / 3600.0
    distance_mi = 10.0 / 1.61
    
    avg_time_per_mi = time_hr / distance_mi
    avg_speed_mph = distance_mi / time_hr
    print avg_time_per_mi #hours per mile
    print avg_speed_mph   #miles per hour
python_calc()

0.116725
8.56714499893


## Chapter 2: Variables, Expression, Statements

**values** belong to different **types**
* integer (ex: 2)
* string (ex: "Hi")
* float/floating point (ex: 3.2)

how to identify what type a value has:

| input           | output       |
| --------------- |:------------:|
| type('example') | type 'str'   |

In [8]:
type('example')

str

**assignment statement** creates new variables and assigns them values
* variable name must start with lowercase letter
  * no illegal characters
  * can't be a keyword
  
  Python 2 keywords: and, del, from, not, as, elif, global, or, while, assert, else, if, pass, with, break, except, import, print, yield, class, exec, in, raise, continue, finally, is, return, def, for, lamba, try

**OPERATORS** = + | - | * | / | **
   operands = values operator is applied to
   floor division -- int/int = int 
    * chops off decimal
    * at least 1 integer must be float for the answer to be a float
   order of operations = PEMDAS
    * only can use (+) or (*) with strings
    * (+) = concatenation (see below)

In [9]:
'I' + 'S'

'IS'

    * (*) = repetition

In [10]:
'I' * 3

'III'

### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1) 8 | integer

2) 8.5 | float

3) 4.0 | float

4) 11 | integer

5) '.....' | string

### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [14]:
import math
def vol_sphere():
    radius = 5
    volume = (4.0/3.0) * (math.pi) * (radius**3)
    print volume
vol_sphere()

523.598775598


Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [17]:
def bookstore():
    cover_price = 24.95 #dollars
    discount = 0.6
    shipping = 0.75 #dollars
    
    total = 3.0 + 60 * cover_price * discount + 59 * shipping
    print total
bookstore()

945.45


If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [23]:
def breakfast_time():
    running_time = 2.0 * 8.25 + 3.0 * 7.2 #minutes
    time = running_time - 8
    print '7:' + str(time) + ' am'
breakfast_time()
    

7:30.1 am


## Chapter 3: Functions

function(argument)

  * ex) type() | int() | float() | str()

**module** = file with collection of related functions
  * MUST IMPORT by:
    * from module import function
      * from math import sin
    * module.function()
      * "dot notation" 
      * math.sin()

**function definition** = create new function

In [24]:
def function_name():  #this is the header
    print 3           #this is the body

**flow of execution** = top to bottom, one statement at a time
  * must call function to execute body

Argument is assigned to parameter = variable
variables and parameters created inside the function are **local**
  * local = exists only in function

**stack diagram** shows value of each variable *and* function each variable belongs to. Example below:

![alt text](https://pythonproject.files.wordpress.com/2014/02/exercise_4-1-2_stack_diagram.jpg)

**fruitful function** yield results
  * **void function** doesn't return values

Can import everything in module using:

In [25]:
from math import *

### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
                                                                 allen```

In [28]:
def right_justify(x):
    print ' ' * 70 + x
right_justify('s')

                                                                      s


### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [42]:
def grid():
    def row():
        return '+ ' + '- '*4 + '+ ' + '- '*4 + '+\n' 
    def column():
        return '|' + ' '*9 + '|' + ' '*9 + '|\n'
    print row() + column() * 4 + row() + column() * 4 + row()
grid()


+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +



(b) Write a function that draws a similar grid with four rows and four columns.

In [47]:
def grid():
    def row():
        return ('+ ' + '- '*4 + '+ ' + '- '*4)*2 + '+\n' 
    def column():
        return ('|' + ' '*9 + '|' + ' '*9)*2 + '|\n'
    print (row() + column() * 4)*4 + row()
grid()


+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +



## Chapter 5.1 - 5.7: Conditionals and Recursion

**modulus operator** gives remainder of integers

In [48]:
remainder = 7%3
print remainder

1


**boolean** = True/False expression
  * T + F != strings | their type = bool

RELATIONAL OPERATORS


|RO      |What it means            |RO        |What it means        |
| ------ |:-----------------------:|: -------:| -------------------:|
| x == y | equal to                | x > y    | greater than        |
| x != y | not equal to            | x < y    | less than           |
| x >= y | greater than/equal to   | x <= y   | less than/equal to  |

* Logical Operators = and, or not

* **Conditional statements** = if/then statements
  * if true, function will execute
* **Alternative execution** = if/else
  * alternatives = **branches**
* **chained conditionals** = if/elif/else
* **nestetd conditionals**

**recursive** function calls itself (this is called **recursion**
  * infinite recursion is **_bad_**, leads to ERROR

/n = **newline** = line break

### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [52]:
def check_fermat(a,b,c,n):
    if n > 2 and a**n + b**n == c**n:
        print "Holy schnitzel, Fermat was wrong!"
    else:
        print "No, that doesn't work"
check_fermat(2,2,2,5)
check_fermat(2,3,6,1)
check_fermat(4,3,2,10)

No, that doesn't work
No, that doesn't work
No, that doesn't work


(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [54]:
def check_fermat(a,b,c,n):
    if n > 2 and a**n + b**n == c**n:
        print "Holy schnitzel, Fermat was wrong!"
    else:
        print "No, that doesn't work"
def integerize(a,b,c,n):
    a = int(a)
    b = int(b)
    c = int(c)
    n = int(n)
    check_fermat(a,b,c,n)
integerize(2,2,2,5)
integerize(2,3,6,1)
integerize(2.0,3.5,6,2)

No, that doesn't work
No, that doesn't work
No, that doesn't work


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [55]:
def is_triangle(a,b,c):
    if a+b<c or a+c<b or b+c<a:
        print "No"
    else:
        print "Yes"
is_triangle(1,3,2)
is_triangle(5,6,20)

Yes
No


(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [56]:
def is_triangle(a,b,c):
    if a+b<c or a+c<b or b+c<a:
        print "No"
    else:
        print "Yes"
def integerize(a,b,c):
    a = int(a)
    b = int(b)
    c = int(c)
    is_triangle(a,b,c)
integerize(1,3,2)
integerize(5,6,20)
integerize(1.3,3,2)

Yes
No
Yes


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.